ROADMAP.md

```diff
@@ -30,41 +30,212 @@ implementation.
+data requested for deletion should be removed and no other data. Removing
+unintended data is worse than _not_ removing data that was requested for
+removal but ideally, both are supported. Generally, according to this rule, we
+err on holding data longer than needed, ensuring that it is only removed when
+we can be certain that it can be removed. With the current behavior, we opt to
+hold onto the data forever, ensuring that data cannot be incorrectly removed.
+
+To understand the problems with implementing deletes, one must understand the
+data model. All registry data is stored in a filesystem layout, implemented on
+a "storage driver", effectively a virtual file system. The storage system must
+assume that this VFS layer will be eventually consistent and has poor read-
+after-write consistency, since this is the lower common denominator among the
+storage drivers. This is mitigated by writing values in reverse-dependent
+order, but makes wider transactional operations unsafe.
+
+Layered on the VFS model is a content-addressable DAG made up of blobs.
+Manifests reference layers. Tags reference manifests. Since the same data can
+be referenced by multiple manifests, we only store data once, even if it is in
+different repositories. Thus, we have a set of blobs, referenced by tags and
+manifests. If we want to delete a blob we need to be certain that it is no
+longer referenced by another manifest or tag. When we delete a manifest, we
+also can try to delete the referenced blobs. Deciding whether or not a blob
+has an active reference is the crux of the problem.
+
+Conceptually, deleting a manifest and its resources is quite simple. Just find
+all the manifests, enumerate the referenced blobs and delete the blobs not in
+that set. An astute observer will recognize this as a garbage collection
+problem. As with garbage collection in programming languages, this is very
+simple when one always has a consistent view. When one adds parallelism and an
+inconsistent view of data, it becomes very challenging.
+
+A simple example can demonstrate this. Let's say we are deleting a manifest
+_A_ in one process. We scan the manifest and decide that all the blobs are
+ready for deletion. Concurrently, we have another process accepting a new
+manifest _B_ referencing one or more blobs from the manifest _A_. Manifest _B_
+is accepted and all the blobs are considered present, so the operation
+proceeds. The original process then deletes the referenced blobs, assuming
+they were unreferenced. The manifest _B_, which we thought had all of its data
+present, can no longer be served by the registry, since the dependent data has
+been deleted.
+
+Deleting data from the registry safely requires some way to coordinate this
+operation. The following approaches are being considered:
+
+- _Reference Counting_ - Maintain a count of references to each blob. This is
+  problematic since we have to do the same GC scanning to build up the
```

Comment:
I don't think this has to be problematic.  For a new registry deployment there won't be any scanning required - just reference increment/decrement on manifest push / delete.  For an existing deployment, this is a one off on registry startup

The crux is areference count consensus amongst a pool of registries sharing one backend, which something like Raft should be able to deal with.